LibertyDSNP/sdk-ts
# Feature/fetch logs paged

Problem
=======
We want the ability to fetch Batch Publications asychronously and lazily loaded.
[Finishes #179436295](https://www.pivotaltracker.com/story/show/179436295)

Solution
========
* Create a class AsyncPublicationsIterator that conforms to AsyncIterator (but only in that it has an async next() function)
* Create a function that returns the new iterator after doing some range checking.
* Tests for same

How `next()` works:
1. Calling `next()` on the iterator makes it search for filtered logs on chain, `walkbackBlocks` at a time until it gets results.  
2. Once it gets at least one result it, returns an "IteratorResult" which is an object containing a "done" boolean meaning no more results are left, and the next Publication result, or undefined if there are none.
3. Subsequent calls to `next()` advance the index into the fetched publications until it runs out.  
4. When next has returned the last of the fetched results, calling `next()` repeats the process of walking back until it gets results, or until the `earliestBlock` value has been reached, whichever comes first.

Double Checks:
---------------
- [] Did you update the changelog?
- [X] Any new modules need to be exported? N/A
- [X] Are new methods in the right module?  I think so?
- [X] Are new methods/modules in core or not (i.e. porcelain or plumbing)?
- [X] Do you have good documentation on exported methods?

Change summary:
---------------
* New class, AsyncPublicationsIterator
* Creator function for above
* Tests for same

Steps to Verify:
----------------
1. Tests should all pass.
1. You should be able to create and use the AsyncPublicationsIterator when you import this version of the sdk.
